src/components/RoomNameInput.js

```diff
@@ -0,0 +1,156 @@
+
+    ...withLocalizePropTypes,
+    ...fullPolicyPropTypes,
+
+    /* Onyx Props */
+
+    /** All reports shared with the user */
+    reports: PropTypes.shape({
+        /** The report name */
+        reportName: PropTypes.string,
+
+        /** ID of the report */
+        reportID: PropTypes.number,
+    }).isRequired,
+
+    /** The policies which the user has access to and which the report could be tied to */
+    policies: PropTypes.shape({
+        /** The policy name */
+        name: PropTypes.string,
+
+        /** ID of the policy */
+        id: PropTypes.string,
+    }).isRequired,
+
+};
+
+const defaultProps = {
+    onChangeText: () => {},
+    onChangeError: () => {},
+    initialValue: '',
+    disabled: false,
+    policyID: '',
+
+    ...fullPolicyDefaultProps,
+};
+
+
+class RoomNameInput extends Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            roomName: props.initialValue,
+            originalRoomName: props.initialValue,
+            error: '',
+        };
+
+        this.checkAndModifyRoomName = this.checkAndModifyRoomName.bind(this);
+    }
+
+    componentDidUpdate(prevProps, prevState) {
+        // As we are modifying the text input, we'll bubble up any changes/errors so the parent component can see it
+        if (prevState.roomName !== this.state.roomName) {
+            this.props.onChangeText(this.state.roomName);
+        }
+        if (prevState.error !== this.state.error) {
+            this.props.onChangeError(this.state.error);
+        }
+    }
+
+    /**
+     * Modifies the room name to follow our conventions:
+     * - Max length 80 characters
+     * - Cannot not include space or special characters, and we automatically apply an underscore for spaces
+     * - Must be lowercase
+     * Also checks to see if this room name already exists, and displays an error message if so.
+     * @param {String} roomName
+     *
+     */
+    checkAndModifyRoomName(roomName) {
+        const modifiedRoomNameWithoutHash = roomName.substring(1)
+            .replace(/ /g, '_')
+            .replace(/[^a-zA-Z\d_]/g, '')
+            .substring(0, CONST.REPORT.MAX_ROOM_NAME_LENGTH)
+            .toLowerCase();
+        const finalRoomName = `#${modifiedRoomNameWithoutHash}`;
+
+        const isExistingRoomName = _.some(
+            _.values(this.props.reports),
+            report => report && report.policyID === this.props.policyID && report.reportName === finalRoomName,
+        );
+
+        let error = '';
+
+        // We error if the room name already exists. We don't care if it matches the original name provided in this
+        // component because then we are not changing the room's name.
+        if (isExistingRoomName && finalRoomName !== this.state.originalRoomName) {
+            error = this.props.translate('newRoomPage.roomAlreadyExists');
+        }
+
+        this.setState({
+            roomName: finalRoomName,
+            error,
+        });
+    }
```

Comment:
Let's just use a ternary here?

```suggestion
        // We error if the room name already exists. We don't care if it matches the original name provided in this
        // component because then we are not changing the room's name.
        const error = isExistingRoomName && finalRoomName !== this.state.originalRoomName ? this.props.translate('newRoomPage.roomAlreadyExists') : '';
        this.setState({
            roomName: finalRoomName,
            error,
        });
    }
```

Reply:
Unless you think the line is too long, honestly NAB

Reply:
Nah this is cleaner lets do it this way.